File: pkg/earthsci-ast-ts/src/brownian.test.ts
/**
 * Brownian (SDE) support — see tests/fixtures/sde/*.
 *
 * From esm 1.0.0 there is no `brownian` variable TYPE. A noise source is a
 * PARAMETER carrying a `distribution` plus `update: {kind: "wiener"}`: the
 * distribution is what gets resampled, every step, with √dt increment scaling.
 * Whether a parameter is Brownian is therefore DERIVED (`brownianParameters`,
 * esm-spec §6.3.1), and correlated noise is one vector-valued parameter whose
 * distribution carries a `cov` matrix rather than the opaque `correlation_group`
 * tag the 0.x schema used and never supplied a matrix for.
 */
import { describe, it, expect } from 'vitest'
import { loadString } from './parse.js'
import { toJson } from './serialize.js'
import { flatten } from './flatten.js'
import { brownianParameters, parameterClass, systemKind } from './classification.js'
import { readFixture } from './test-helpers.js'
import type { Model } from './types.js'

describe('Brownian (SDE) support', () => {
  it('round-trips the Ornstein–Uhlenbeck fixture preserving the wiener parameter', () => {
    const fixture = readFixture('fixtures', 'sde', 'ornstein_uhlenbeck.esm')
    const parsed = loadString(fixture)
    const model = parsed.models!.OU as Model
    const bw = model.variables.Bw

    // Declared as a plain parameter; its Brownian-ness is derived.
    expect(bw.type).toBe('parameter')
    expect(bw.distribution).toEqual({ kind: 'normal', mean: 0.0, std: 1.0 })
    expect(bw.update).toEqual({ kind: 'wiener' })
    expect(parameterClass(bw)).toBe('brownian')
    expect(brownianParameters(model)).toEqual(['Bw'])

    // One wiener parameter is what makes the enclosing model an SDE.
    expect(systemKind(model)).toBe('sde')

    const out = toJson(parsed)
    const reparsed = loadString(out)
    expect((reparsed.models!.OU as Model).variables.Bw).toEqual(bw)
  })

  it('does not mistake the other parameters for noise sources', () => {
    const fixture = readFixture('fixtures', 'sde', 'ornstein_uhlenbeck.esm')
    const model = loadString(fixture).models!.OU as Model

    // `sigma` carries the noise AMPLITUDE and is an ordinary constant; only the
    // parameter with the `wiener` update is Brownian.
    expect(parameterClass(model.variables.sigma)).toBe('constant')
    expect(parameterClass(model.variables.theta)).toBe('constant')
  })

  it('flatten surfaces brownian parameters as a SUBSET of the parameters', () => {
    const fixture = readFixture('fixtures', 'sde', 'correlated_noise.esm')
    const parsed = loadString(fixture)
    const flat = flatten(parsed)
    // esm-spec §6.3.1: the four parameter sets PARTITION the parameters, so the
    // wiener entry is in BOTH maps. `brownianParameters` was `brownianVariables`
    // before 1.0.0, and it EXCLUDED the entry from `parameters` — which made the
    // parameter vector's length depend on whether the model was stochastic.
    expect(Object.keys(flat.brownianParameters)).toEqual(['TwoBody.B'])
    expect(Object.keys(flat.parameters)).toContain('TwoBody.B')
    // Carrying the bucket is exactly what lets the flattened form report "sde".
    expect(flat.systemKind).toBe('sde')
  })

  it('schema rejects a wiener update on an UNKNOWN', () => {
    // `distribution` and `update` are parameter-only: an unknown's behaviour is
    // stated by the equations and nowhere else.
    const bad = JSON.stringify({
      esm: '1.0.0',
      metadata: { name: 'Bad' },
      models: {
        M: {
          variables: {
            x: {
              type: 'unknown',
              units: '1',
              distribution: { kind: 'normal', mean: 0.0, std: 1.0 },
              update: { kind: 'wiener' },
            },
          },
          equations: [],
        },
      },
    })
    expect(() => loadString(bad)).toThrow()
  })

  it('schema rejects a wiener update with no distribution to resample', () => {
    const bad = JSON.stringify({
      esm: '1.0.0',
      metadata: { name: 'Bad' },
      models: {
        M: {
          variables: {
            x: { type: 'unknown', units: '1' },
            w: { type: 'parameter', units: '1/s^0.5', update: { kind: 'wiener' } },
          },
          equations: [{ lhs: { op: 'D', args: ['x'], wrt: 't' }, rhs: 'w' }],
        },
      },
    })
    expect(() => loadString(bad)).toThrow()
  })
})

File: pkg/earthsci-ast-go/pkg/esm/brownian_test.go
package esm

import (
	"encoding/json"
	"os"
	"path/filepath"
	"reflect"
	"testing"
)

// brownian_test.go covers the esm 1.0.0 spelling of a stochastic noise source.
// There is no `brownian` variable TYPE any more: a noise source is a PARAMETER
// carrying a `distribution` and `update: {kind: "wiener"}`, and "which
// parameters are Brownian" is answered by the classification function
// BrownianParameters (esm-spec §6.3.1), never by reading a declared type.
// Correlated noise is likewise one VECTOR-VALUED parameter whose distribution
// carries an explicit `cov`, replacing the 0.x `correlation_group` tag that
// named a correlation without ever giving one.

func loadSDEFixture(t *testing.T, name string) *ESMFile {
	t.Helper()
	fixture := filepath.Join("..", "..", "..", "..", "tests", "fixtures", "sde", name)
	raw, err := os.ReadFile(fixture)
	if err != nil {
		t.Fatalf("read fixture: %v", err)
	}
	var parsed ESMFile
	if err := json.Unmarshal(raw, &parsed); err != nil {
		t.Fatalf("unmarshal fixture: %v", err)
	}
	return &parsed
}

// A wiener-updated parameter round-trips through parse → serialize → parse with
// its distribution and its update rule intact.
func TestWienerParameterRoundTrip(t *testing.T) {
	parsed := loadSDEFixture(t, "ornstein_uhlenbeck.esm")
	bw, ok := parsed.Models["OU"].Variables["Bw"]
	if !ok {
		t.Fatalf("Bw variable missing in fixture")
	}
	if bw.Type != VarTypeParameter {
		t.Errorf("Bw.Type = %q, want %q — a noise source is a parameter in 1.0.0", bw.Type, VarTypeParameter)
	}
	rules := bw.UpdateRules()
	if len(rules) != 1 || rules[0].Kind != UpdateKindWiener {
		t.Fatalf("Bw update = %+v, want a single wiener rule", rules)
	}
	if bw.Distribution == nil || bw.Distribution.Kind != DistributionNormal {
		t.Fatalf("Bw.Distribution = %+v, want a normal distribution (a wiener update resamples it)", bw.Distribution)
	}

	out, err := json.Marshal(parsed)
	if err != nil {
		t.Fatalf("marshal: %v", err)
	}
	var reparsed ESMFile
	if err := json.Unmarshal(out, &reparsed); err != nil {
		t.Fatalf("unmarshal round-trip: %v", err)
	}
	rbw := reparsed.Models["OU"].Variables["Bw"]
	if !reflect.DeepEqual(bw, rbw) {
		t.Errorf("wiener parameter lost information on round-trip: got %+v want %+v", rbw, bw)
	}
}

// The four parameter sets of esm-spec §6.3.1 partition, and only the
// wiener-updated parameter is Brownian. `sigma` — the noise INTENSITY — is an
// ordinary constant, which is the mistake this pins against: a binding that
// treated "anything with noise-ish units" or "anything with a distribution" as
// Brownian would put it in the wrong set.
func TestBrownianParametersClassification(t *testing.T) {
	parsed := loadSDEFixture(t, "ornstein_uhlenbeck.esm")
	model := parsed.Models["OU"]

	if got, want := BrownianParameters(&model), []string{"Bw"}; !reflect.DeepEqual(got, want) {
		t.Errorf("BrownianParameters = %v, want %v", got, want)
	}
	if got, want := ConstantParameters(&model), []string{"sigma", "theta"}; !reflect.DeepEqual(got, want) {
		t.Errorf("ConstantParameters = %v, want %v", got, want)
	}
	if got := DiscreteParameters(&model); len(got) != 0 {
		t.Errorf("DiscreteParameters = %v, want none", got)
	}
	if got := SampledParameters(&model); len(got) != 0 {
		t.Errorf("SampledParameters = %v, want none — Bw has an update, so it is Brownian, not sampled", got)
	}
	// Any Brownian parameter promotes the model to an SDE.
	if got := SystemKind(&model, parsed.Domain); got != SystemKindSDE {
		t.Errorf("SystemKind = %q, want %q", got, SystemKindSDE)
	}
}

// Correlated noise: ONE vector-valued parameter whose distribution carries the
// covariance matrix. The `cov` must survive the round-trip — it is the only
// place the correlation is stated.
func TestCorrelatedNoiseIsOneVectorParameter(t *testing.T) {
	parsed := loadSDEFixture(t, "correlated_noise.esm")
	model := parsed.Models["TwoBody"]

	if got, want := BrownianParameters(&model), []string{"B"}; !reflect.DeepEqual(got, want) {
		t.Errorf("BrownianParameters = %v, want %v (one vector-valued source, not two scalars)", got, want)
	}
	b := model.Variables["B"]
	if b.Distribution == nil {
		t.Fatal("B carries no distribution")
	}
	wantCov := [][]float64{{1.0, 0.5}, {0.5, 1.0}}
	if !reflect.DeepEqual(b.Distribution.Cov, wantCov) {
		t.Errorf("B.Distribution.Cov = %v, want %v", b.Distribution.Cov, wantCov)
	}
	if got, want := b.Dims(), []string{"wind_noise"}; !reflect.DeepEqual(got, want) {
		t.Errorf("B shape = %v, want %v (the cov order must match it)", got, want)
	}

	out, err := json.Marshal(parsed)
	if err != nil {
		t.Fatalf("marshal: %v", err)
	}
	var reparsed ESMFile
	if err := json.Unmarshal(out, &reparsed); err != nil {
		t.Fatalf("unmarshal round-trip: %v", err)
	}
	if !reflect.DeepEqual(b, reparsed.Models["TwoBody"].Variables["B"]) {
		t.Errorf("correlated-noise parameter lost information on round-trip: got %+v want %+v",
			reparsed.Models["TwoBody"].Variables["B"], b)
	}
}

// Flattening surfaces the derived Brownian set under its dot-namespaced names.
// The parameter is ALSO an ordinary parameter of the flattened system, which is
// the substantive change from 0.x: a noise source is no longer a variable kind
// of its own.
func TestFlattenBrownianParameters(t *testing.T) {
	parsed := loadSDEFixture(t, "correlated_noise.esm")
	flat, err := Flatten(parsed)
	if err != nil {
		t.Fatalf("flatten: %v", err)
	}
	names := func(vs []FlattenedVariable) []string {
		out := []string{}
		for _, v := range vs {
			out = append(out, v.Name)
		}
		return out
	}
	if want := []string{"TwoBody.B"}; !reflect.DeepEqual(names(flat.BrownianParameters), want) {
		t.Errorf("BrownianParameters = %v, want %v", names(flat.BrownianParameters), want)
	}
	if !containsVar(flat.Parameters, "TwoBody.B") {
		t.Errorf("Parameters = %v, want it to include TwoBody.B — a noise source IS a parameter",
			names(flat.Parameters))
	}
	// The Brownian bucket carries the FULL variable, not a bare name: a consumer
	// building an SDE problem needs its distribution and shape from here
	// (esm-libraries-spec §4.7.5 step 4, "Full metadata, not names").
	if b := flat.BrownianParameters[0]; b.Distribution == nil || len(b.Shape) == 0 {
		t.Errorf("flattened Brownian parameter lost its metadata: %+v", b)
	}
	if want := []string{"TwoBody.x", "TwoBody.y"}; !reflect.DeepEqual(names(flat.StateVariables), want) {
		t.Errorf("StateVariables = %v, want %v", names(flat.StateVariables), want)
	}
}
